File: matlabm/bankfit/spectrafit.m
function [ra] = spectrafit(C,d,x0,options)

tol = 10*eps*norm(C,1)*length(C);
[m,n] = size(C);
P = zeros(1,n);
Z = 1:n;
ra = P';

ZZ=Z;
resid = d-C*ra;
w = C'*(resid);


% set up iteration criterion
outeriter = 0;
iter = 0;
itmax = 3*n;
exitflag = 1;
kkk = 0;
while any(Z) & any(w(ZZ) > tol)
   
   www = w(ZZ);
    outeriter = outeriter + 1;
   [wt,t] = max(w(ZZ));
   t = ZZ(t);
   P(1,t) = t;
   Z(t) = 0;
   PP = find(P);
   ZZ = find(Z);
   nzz = size(ZZ);
   CP(1:m,PP) = C(:,PP);
   CP(:,ZZ) = zeros(m,nzz(2));
   z = pinv(CP)*d;
   z(ZZ) = zeros(nzz(2),nzz(1));
   % inner loop to remove elements from the positive set which no longer belong
   
   while any((z(PP) <= tol))
      ttt = z(PP);
       iter = iter + 1;
      if iter > itmax
         
         exitflag = 0;
         output.iterations = outeriter;
         resnorm = sum(resid.*resid);
         ra = z;
         lambda = w;
         return
      end
      QQ = find((z <= tol) & P');
      alpha = min(ra(QQ)./(ra(QQ) - z(QQ)));
      ra = ra + alpha*(z - ra);
      ij = find(abs(ra) < tol & P' ~= 0);
      Z(ij)=ij';
      P(ij)=zeros(1,length(ij));
      PP = find(P);
      ZZ = find(Z);
      nzz = size(ZZ);
      CP(1:m,PP) = C(:,PP);
      CP(:,ZZ) = zeros(m,nzz(2));
      z = pinv(CP)*d;
      z(ZZ) = zeros(nzz(2),nzz(1));
   end
   ra = z;
   resid = d-C*ra;
   w = C'*(resid);
   kkk = kkk+1;
end

File: RIPP.Lib/MathLib/Fit/SpectraFit.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using MathNet.Numerics.LinearAlgebra.Double;
using RIPP.Lib.MathLib;

namespace RIPP.Lib.MathLib.Fit
{
    public class SpectraFit
    {
        public static Vector Fit(Matrix input, Vector y)
        {
            if (input == null || y == null)
                throw new ArgumentNullException("输入不能为空");
            if (input.RowCount != y.Count)
                throw new ArgumentException("输入的长度不一致");
            //tol = 10*eps*norm(C,1)*length(C);
            // norm(C,1) = sum(abs(C))
            var norm = new DenseVector(input.ColumnCount);
            for (int i = 0; i < input.RowCount; i++)
                for (int j = 0; j < input.ColumnCount; j++)
                    norm[j] += Math.Abs(input[i, j]);

            var tol = 10 * Single.Epsilon * norm.Sum() * input.RowCount;


            var P = new DenseVector(input.ColumnCount);
            var Z = new DenseVector(input.ColumnCount);
            for (int i = 1; i <= Z.Count; i++)
                Z[i-1] = i;
            var ra = P.Clone();
            var ZZ = Z.Clone();
            var resid = y - input * ra;
            var w = input.Transpose() * resid;


            // set up iteration criterion
            var outeriter = 0;
            var iter = 0;
            var itmax = 3 * input.ColumnCount;
            var exitflag = 1;


            while (Z.Any())
            {
                var widx = ((Vector)w).Slice((Vector)ZZ.Subtract(1));
                bool brk = true;
                foreach (var www in widx)
                    if ((www - tol) > 0)
                    {
                        brk = false;
                        break;
                    }
                if (brk)
                    break;

                if (!widx.Any())
                    break;
                outeriter++;
                var wt = widx.Maximum();
                var t = widx.MaximumIndex();
                t = (int)ZZ[t];
                P[t-1] = t;
                Z[t-1] = 0;
                var PP = P.Find();
                ZZ = Z.Find().Add(1);
                var nzz = ZZ.Count;

                var CP = new DenseMatrix(input.RowCount, input.ColumnCount);

                for (int j = 0; j < input.RowCount; j++)
                {
                    foreach (var pj in PP)
                        CP[j, (int)pj] = input[j, (int)pj];
                    foreach (var zj in ZZ)
                        CP[j, (int)zj-1] = 0;
                }

                var z = CP.PseudoInverse() * y;
                foreach (var zj in ZZ)
                    z[(int)zj-1] = 0;


                while (true)
                {
                    var ztemp = ((Vector)z).Slice((Vector)PP);
                    brk = true;
                    foreach (var ztempp in ztemp)
                        if (ztempp <= tol)
                        {
                            brk = false;
                            break;
                        }
                    if (brk)
                        break;
                    iter = iter + 1;
                    if (iter > itmax)
                    {
                        exitflag = 0;
                        ra = z;
                        break;
                    }

                    var lst = new List<int>();
                    for (int i = 0; i < z.Count; i++)
                    {
                        if (z[i] <= tol && P[i] != 0)
                            lst.Add(i);
                    }
                    var QQ = lst.ToArray();
                    var ratemp = ((Vector)ra).Slice(QQ);
                    var alpha = (ratemp.PointwiseDivide(ratemp - ((Vector)z).Slice(QQ))).Min();
                    ra = ra + alpha * (z - ra);
                    lst = new List<int>();
                    for (int i = 0; i < ra.Count; i++)
                    {
                        if (Math.Abs(ra[i]) < tol && P[i] != 0)
                            lst.Add(i);
                    }
                    foreach (var i in lst)
                    {
                        Z[i] = i;
                        P[i] = 0;
                    }
                    PP = P.Find();
                    ZZ = Z.Find().Add(1);
                    nzz = ZZ.Count;
                    for (int j = 0; j < input.RowCount; j++)
                    {
                        foreach (var pj in PP)
                            CP[j, (int)pj] = input[j, (int)pj];
                        foreach (var zj in ZZ)
                            CP[j, (int)zj-1] = 0;
                    }

                    z = CP.PseudoInverse() * y;
                    foreach (var zj in ZZ)
                        z[(int)zj-1] = 0;


                }
                ra = z;
                resid = y - input * ra;
                w = input.Transpose() * resid;

            }

            return (Vector)ra;

        }




    }
}
